File: coding/solutions/src/main/java/xyz/grind/coding/blind75/ValidSudoku.java
package xyz.grind.coding.blind75;

import java.util.Arrays;
import java.util.HashSet;

public class ValidSudoku {

    public static boolean isValidSudoku(char[][] board){

        HashSet<String> seen = new HashSet<String>();
        for(int row = 0; row < board.length; row++) {
            for(int column = 0; column < board.length; column++) {
                char currentVal = board[row][column];
                if(currentVal == '.') continue;
                if(!seen.add(currentVal + " Found in row " + row) ||
                   !seen.add(currentVal + " Found in column " + column) ||
                   !seen.add(currentVal + " Found in sub-box " + row/3 + " - " + column/3)) return false;
                }
            }
        return true;
    }

    public static boolean isValidSudokuHashSet(char[][] board) {

        int boardLength = 9;
        HashSet<Character>[] rows = new HashSet[boardLength];
        HashSet<Character>[] columns = new HashSet[boardLength];
        HashSet<Character>[] boxes = new HashSet[boardLength];

        for(int i = 0; i < boardLength; i++) {
            rows[i] = new HashSet<>();
            columns[i] = new HashSet<>();
            boxes[i] = new HashSet<>();
        }

        for(int r = 0; r < boardLength; r++) {
            for(int c = 0; c < boardLength; c++) {
                char currentValue = board[r][c];
                if(currentValue == '.') continue;

                if(rows[r].contains(currentValue)) return false;
                rows[r].add(currentValue);

                if(columns[c].contains(currentValue)) return false;
                columns[c].add(currentValue);

                // this tells us the current square we're in
                int boxIndex = (r / 3) * 3 + (c / 3);
                if(boxes[boxIndex].contains(currentValue)) return false;
                boxes[boxIndex].add(currentValue);

            }
        }

        return true;

    }

}

File: problem-solving/solutions/leetcode/blind75/src/main/kotlin/xyz/grind/coding/easy/ValidSudoku.kt
package xyz.grind.coding.easy

object ValidSudoku {

    // Commented to Improve test coverage for the solutions that matter
    //    @JvmStatic
    //    fun main(args: Array<String>) {
    //
    //        // Experimenting with the problem's data structures
    //        val sudokuBoard = arrayOf(
    //            charArrayOf('5', '3', '.', '.', '7', '.', '.', '.', '.'),
    //            charArrayOf('6', '.', '.', '1', '9', '5', '.', '.', '.'),
    //            charArrayOf('.', '9', '8', '.', '.', '.', '.', '6', '.'),
    //            charArrayOf('8', '.', '.', '.', '6', '.', '.', '.', '3'),
    //            charArrayOf('4', '.', '.', '8', '.', '3', '.', '.', '1'),
    //            charArrayOf('7', '.', '.', '.', '2', '.', '.', '.', '6'),
    //            charArrayOf('.', '6', '.', '.', '.', '.', '2', '8', '.'),
    //            charArrayOf('.', '.', '.', '4', '1', '9', '.', '.', '5'),
    //            charArrayOf('.', '.', '.', '.', '8', '.', '.', '7', '9')
    //        )
    //        val set: Array<HashSet<String>> = arrayOf(HashSet(9))
    //        val setTwo: Array<HashSet<String>> = arrayOf(HashSet(9))
    //        for (i in set.indices) {
    //            set[i] = HashSet()
    //            setTwo[i] = HashSet()
    //        }
    //        for (i in set.indices) {
    //            for (j in set.indices) {
    //                set[i].add("Something")
    //                setTwo[i].add("Something")
    //            }
    //        }
    //        println(set.contentToString())
    //        println(setTwo.contentToString())
    //        println(isValidSudokuHashSet(sudokuBoard))
    //    }

    fun isValidSudoku(board: Array<CharArray>): Boolean {
        val seen = HashSet<String>()
        for (row in board.indices) {
            for (column in board.indices) {
                val currentVal = board[row][column]
                if (currentVal == '.') continue
                if (!seen.add("$currentVal Found in row $row") ||
                    !seen.add("$currentVal Found in column $column") ||
                    !seen.add(currentVal.toString() + " Found in sub-box " + row / 3 + " - " + column / 3)
                ) return false
            }
        }
        return true
    }

    fun isValidSudokuHashSet(board: Array<CharArray>): Boolean {
        val rows = Array(9) { HashSet<Char>() }
        val columns = Array(9) { HashSet<Char>() }
        val boxes = Array(9) { HashSet<Char>() }

        for (r in 0 until 9) {
            for (c in 0 until 9) {
                val currentValue = board[r][c]
                // this tells us the current square we're in
                val boxIndex = (r / 3) * 3 + (c / 3)
                if (currentValue == '.') continue
                if (rows[r].contains(currentValue) or
                    columns[c].contains(currentValue) or
                    boxes[boxIndex].contains(currentValue)
                ) return false
                rows[r].add(currentValue)
                columns[c].add(currentValue)
                boxes[boxIndex].add(currentValue)
            }
        }

        return true
    }
}
